src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -35,6 +35,518 @@ extern "C" {
+//		function.
+//---------------------------------------------------------------------------
+BOOL
+CQueryMutators::GroupingFuncRewriteWalker(
+	Node *node, SContexGroupingFuncRewriteWalker *context)
+{
+	if (NULL == node)
+	{
+		return false;
+	}
+
+	if (IsA(node, GroupingFunc))
+	{
+		GroupingFunc *gf = (GroupingFunc *) node;
+		ListCell *arg_lc;
+		List *newargs = NIL;
+
+		gf->ngrpcols = context->m_ngrpcols;
+
+		// For each argument in gf->args, find its new position
+		ForEach(arg_lc, gf->args)
+		{
+			int arg = intVal(lfirst(arg_lc));
+			int new_arg =
+				gpdb::ListNthInt(context->m_grouping_tle_refs_mapping, arg);
+			newargs = gpdb::LAppend(newargs, gpdb::MakeIntegerValue(new_arg));
+		}
+
+		// Free gf->args since we do not need it any more.
+		gpdb::ListFreeDeep(gf->args);
+		gf->args = newargs;
+		return false;
+	}
+	return gpdb::WalkExpressionTree(
+		node, (ExprWalkerFn) GroupingFuncRewriteWalker, context);
+}
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::GetVarsWithoutTleWalker
+//
+//	@doc:
+//		Goes over the expression tree, finds all vars that do not have relevant
+//		targetlist entries, and adds resjunk targetlist entries for such vars
+//		to a m_tlist_addition in the context, that later will be added to the
+//		original target list.
+//---------------------------------------------------------------------------
+BOOL
+CQueryMutators::GetVarsWithoutTleWalker(
+	Node *node, SContextFixGroupDependentTargets *context)
+{
+	if (NULL == node)
+	{
+		return false;
+	}
+
+	if (IsA(node, Aggref))
+	{
+		// do not examine what is inside Aggref
+		return false;
+	}
+
+	// find all vars that do not have relevant targetlist entries, and add
+	// resjunk targetlist entries for such vars to m_tlist_addition
+	if (IsA(node, Var))
+	{
+		bool found = false;
+		Var *var = (Var *) gpdb::CopyObject(node);
+		if (var->varlevelsup == context->m_current_query_level)
+		{
+			var->varlevelsup = 0;
+			ListCell *lc_tle = NULL;
+			ForEach(lc_tle, context->m_query->targetList)
+			{
+				TargetEntry *target_entry = (TargetEntry *) lfirst(lc_tle);
+
+				if (IsA(target_entry->expr, Var))
+				{
+					if (gpdb::Equals(target_entry->expr, var))
+					{
+						found = true;
+						break;
+					}
+				}
+			}
+			if (!found)
+			{
+				int ressno = gpdb::ListLength(context->m_query->targetList) + 1;
+				TargetEntry *newTargetEntry =
+					gpdb::MakeTargetEntry((Expr *) var, ressno, NULL, true);
+
+				context->m_tlist_addition =
+					gpdb::LAppend(context->m_tlist_addition, newTargetEntry);
+			}
+		}
+		else
+		{
+			gpdb::GPDBFree(var);
+		}
+
+		return false;
+	}
+
+	if (IsA(node, SubLink))
+	{
+		SubLink *sublink = (SubLink *) node;
+
+		context->m_current_query_level++;
+		GetVarsWithoutTleWalker((Node *) sublink->subselect, context);
+		context->m_current_query_level--;
+
+		return false;
+	}
+
+	// recurse into query structure
+	if (IsA(node, Query))
```

Comment:
In RunExtractAggregatesMutator we have additionally the CommonTableExpr, can it be related to us as well?

Reply:
It looks that CTE is processed separately by CQueryMutators::NormalizeQuery, for ex, if executiong such sample query 
```
WITH cte1 AS (
    SELECT c
    FROM test_table1
    GROUP BY a
)
SELECT b, c
FROM test_table3
WHERE c IN (SELECT cte1.c FROM cte1)
GROUP BY a ORDER BY b, c;
```
I've got the following query tree structures (if added printing of query tree before and after FixGroupDependentTargets) - [cte_log.txt](https://github.com/arenadata/gpdb/files/14358967/cte_log.txt) . The log is quit long, so I've placed it in a separate file. From the log it can be seen that FixGroupDependentTargets is called twice - first for the origianl query, and second for the cte itself. Do you have other examples of queries with CTE that could be problematic?

Reply:
No, I can't find any appropriate example. Let's skip it then, I hope it won't fail accidently.